### PR TITLE
Fixing bugs in icon-magic

### PR DESCRIPTION
**Pull request by ramitha**
- Removing the try catch blocks from the cli so the errors are thrown
- Logging errors for missing variants instead of throwing errors
- Adding outputPath to the configSchema
- Modifying the paths written to the output config file (iconrc.json) to be relative to the output icon directory
  - Updating the iconPath itself to '.'
  - Updating the sourceConfigFile path
- Needed to write the output of build to the output folder so it serves as the input asset for generate
- Fixing bugs when build and generate are run in two separate steps vs. one after the other
  - the generate plugins were using icon.contents instead of icon.getContents() which can be empty in this case
  - create a wrapper around the generate function that builds the iconSet fresh
  - update the config-schema to optionally contain flavors and sourceConfigPaths which are generated during build
  - optionally skip the check for variants as the generate step begins with flavors and not variants
- Build and generate now support importing plugins by name

svg-generate: fixes and optimization
- Fixing svg-generate as it was writing [object] to the file instead of the actual contents
- Adding data-supported-dps to the svg elements
- Removing all other attributes - width, height, viewbox
- Removing any embedded raster images
- Modifying svg-to-raster naming to be <assetname>-<width>x<height>@<resolution>
- Modifying svg-generate to take in nameSizeMapping
- Tweaking svg-generate plugin options for viewBox instead of width and height
- Adding an id to the svg created with svg-generate

Using workerpool to paralellize icon generation
- Created a generate-worker that carries out generate for a single icon
- generate/index is updated to use a workerpool with generate-worker

Replacing debug with winston
- Creating a wrapper around winston to append the fileName to the log message

### CODE REVIEW

File: packages/@icon-magic/generate/src/generate-worker.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import { svgGenerate } from './plugins/svg-generate';
+import { svgToRaster } from './plugins/svg-to-raster';
+
+const LOGGER: Logger = logger('icon-magic:generate:index');
+
+async function generateSingleIcon(iconConfig: IconConfig) {
+  // TODO: this function should take in an instance of Icon but due to an issue
+  // in workerpool, I'm using a workaround where we're having to create this
+  // instance by taking an iconConfig instead.
+  // Update when it is fixed. Refer to generate/index for more details
```
Comment (hagmandan) [inline packages/@icon-magic/generate/src/generate-worker.ts]: is this a fix we need to implement, or one we can wait for

File: packages/@icon-magic/generate/src/index.ts [hunk 2]
```diff
@@ -23,72 +15,45 @@ const debug = debugGenerator('icon-magic:generate:index');
  * After generate has applied all the plugins based on type, we now get flavors
  * with types that contain paths to the newly created .type asset. Generate also
  * updates the icon config with the newly generated types.
+ *
+ * If no plugins are passed for the type, then by default, svgToRaster is
+ * applion raster and svgGenerate is applied on svg types
```
Comment (oa495) [inline packages/@icon-magic/generate/src/index.ts]: nit: `applion` -> `applied`

File: packages/@icon-magic/build/src/index.ts [hunk 3]
```diff
@@ -37,37 +37,43 @@ export async function build(iconConfig: IconConfigHash): Promise<IconSet> {
...
+
     // get the output directory with respect to the current working directory
     // and then create a directory with the iconName
-    const outputPath = icon.getBuildOutputPath();
+    const buildOutputPath = icon.getBuildOutputPath();
+    const iconOutputPath = icon.getIconOutputPath();
 
     // create the directory if it doesn't already exist
-    await fs.mkdirp(outputPath);
+    await fs.mkdirp(buildOutputPath);
```
[the remark this answers is not in the document]
Comment (ramitha) [reply packages/@icon-magic/build/src/index.ts]: So I've changed `icon.ts` a little, so `buildOutputPath` is constructed from `iconOutputPath`. We'll now store the output of `build` within `iconOutputPath` so we can run `build` and `generate` separately. It'll also help for hashing. 

Final output, after build and generate will look like this: 
```
iconOutputPath
├── build
 │   ├── iconrc.json
 │   └── test.svg
├── iconrc.json
├── test-144x144.png
├── test-144x144.webp
├── test-192x192.png
├── test-192x192.webp
├── test-48x48.png
├── test-48x48.webp
├── test-48x48@1.5.png
├── test-48x48@1.5.webp
├── test-48x48@1.png
├── test-48x48@1.webp
├── test-48x48@2.png
├── test-48x48@2.webp
├── test-48x48@3.png
├── test-48x48@3.webp
├── test-48x48@4.png
├── test-48x48@4.webp
├── test-72x72.png
├── test-72x72.webp
├── test-96x96.png
├── test-96x96.webp
└── test.svg
```

File: packages/@icon-magic/build/src/index.ts [hunk 4]
```diff
@@ -37,37 +37,43 @@ export async function build(iconConfig: IconConfigHash): Promise<IconSet> {
...
+      assets.map(asset => saveAssetAsFlavor(asset, icon, buildOutputPath))
     );
 
-    // in the icon, update the iconPath to be that of the output path
-    icon.iconPath = outputPath;
     // add it to the outputIconSet's hash map
-    outputIconSet.hash.set(outputPath, icon);
-    debug(
-      `Icon ${icon.iconName} has been written to the hash as ${icon.iconPath}`
+    outputIconSet.hash.set(iconOutputPath, icon);
```
[the remark this answers is not in the document]
Comment (ramitha) [reply packages/@icon-magic/build/src/index.ts]: Same as above, let's chat about it more in person.

File: packages/@icon-magic/config-reader/src/schemas/config-schema.ts [hunk 5]
```diff
@@ -97,6 +97,9 @@ const topLevelConfigProperties = {
       type: 'number'
     }
   },
+  flavors: { type: 'array' },
```
[the remark this answers is not in the document]
Comment (ramitha) [reply packages/@icon-magic/config-reader/src/schemas/config-schema.ts]: Updated.

**Pull request merged**